crates/rpc/rpc/src/eth/api/fees.rs

```diff
@@ -0,0 +1,189 @@
+        newest_block: BlockId,
+        reward_percentiles: Option<Vec<f64>>,
+    ) -> EthResult<FeeHistory> {
+        let block_count = block_count.as_u64();
+
+        if block_count == 0 {
+            return Ok(FeeHistory::default())
+        }
+
+        let Some(previous_to_end_block) = self.inner.client.block_number_for_id(newest_block)? else { return Err(EthApiError::UnknownBlockNumber)};
+        let end_block = previous_to_end_block + 1;
+
+        if end_block < block_count {
+            return Err(EthApiError::InvalidBlockRange)
+        }
+
+        let mut start_block = end_block - block_count;
+
+        if block_count == 1 {
+            start_block = previous_to_end_block;
+        }
+
+        // if not provided the percentiles are []
+        let reward_percentiles = reward_percentiles.unwrap_or_default();
+
+        // checks for rewardPercentile's sorted-ness
+        // check if any of rewardPercentile is greater than 100
+        // pre 1559 blocks, return 0 for baseFeePerGas
+        for window in reward_percentiles.windows(2) {
+            if window[0] >= window[1] {
+                return Err(EthApiError::InvalidRewardPercentile(window[1]))
+            }
+
+            if window[0] < 0.0 || window[0] > 100.0 {
+                return Err(EthApiError::InvalidRewardPercentile(window[0]))
+            }
+        }
+
+        let mut fee_history_cache = self.fee_history_cache.0.lock().await;
+
+        // Sorted map that's populated in two rounds:
+        // 1. Cache entries until first non-cached block
+        // 2. Database query from the first non-cached block
+        let mut fee_history_cache_items = BTreeMap::new();
+
+        let mut first_non_cached_block = None;
+        let mut last_non_cached_block = None;
+        for block in start_block..=end_block {
+            // Check if block exists in cache, and move it to the head of the list if so
+            if let Some(fee_history_cache_item) = fee_history_cache.get(&block) {
+                fee_history_cache_items.insert(block, fee_history_cache_item.clone());
+            } else {
+                // If block doesn't exist in cache, set it as a first non-cached block to query it
+                // from the database
+                first_non_cached_block.get_or_insert(block);
+                // And last non-cached block, so we could query the database until we reach it
+                last_non_cached_block = Some(block);
+            }
+        }
+
+        // If we had any cache misses, query the database starting with the first non-cached block
+        // and ending with the last
+        if let (Some(start_block), Some(end_block)) =
+            (first_non_cached_block, last_non_cached_block)
+        {
+            let header_range = start_block..=end_block;
+
+            let headers: Vec<Header> = self.inner.client.headers_range(header_range.clone())?;
+            let transactions = self.inner.client.transactions_by_block_range(header_range)?;
```

Comment:
I think it is reasonable to always read them from disk, if we add a service task that inserts items to the cache for every new block we can prevent this for the majority of requests